### ISSUE TITLE
Merit Order charts sometimes take extremely long to load

### ISSUE THREAD

**Issue by dennisschoenmakers** [quintel/etmodel#2117]
I am a bit taken aback by the amount of time it needs to load with the high bandwidth internet connection at our office, also there is no visual cue that something is happening.

![image](https://cloud.githubusercontent.com/assets/104830/15422288/1450ec30-1e78-11e6-87e2-ae73447cdc16.png)

![image](https://cloud.githubusercontent.com/assets/104830/15422302/2ee4736e-1e78-11e6-8dca-91971a4f60e1.png)

cc @jorisberkhout

**Comment by dennisschoenmakers** [comment]
Also, when you move a slider when you have the chart opened, the application is **extremely** slow. It takes >10 seconds to load the page.

cc @antw

**Comment by dennisschoenmakers** [comment]
It seem to be that the request is not slower, it is the browser that is slow. This is even more concerning, because I think my browser (Chrome on a fast laptop) is a **lot** quicker than the average laptop.

**Comment by dennisschoenmakers** [comment]
I am now testing this on a Windows computer and the waiting time for some browser is more than 1 minute.

**Comment by dennisschoenmakers** [comment]
Also, the user interaction on your browser completely freezes. This really is bad, since people cannot croll, or click and they start acting weird.

**Comment by grdw** [comment]
SInce my own machine is pretty specced out, I didn't really notice the slowness. So I'm using a VirtualMachine with the following settings to test a slow machine:

It's a Debian machine (small Linux OS), 1 CPU, 2GB of Memory. On first sight it doesn't feel slow, except for the price curve which is still painfully slow. 

In more detail (for the price curve):
- The average line is the slowest thing in the price curve.

![screen shot 2016-05-20 at 12 08 45](https://cloud.githubusercontent.com/assets/2676542/15424557/a229eb32-1e83-11e6-8d7d-71942239184d.png)

I did found out why the price curve is so slow. The other charts however, am still looking. But they don't feel as slow...

**Comment by dennisschoenmakers** [comment]
Since JavaScript execution in the browser is single threated, we might also consider using timeouts to release execution back to the browser. E.g. http://stackoverflow.com/a/672784. But, first let's optimize our calculations!

**Comment by grdw** [comment]
> But, first let's optimize our calculations!

Yes I agree. The hourly supply chart being the slowest... the main reason is the stacking. It has to stack > 20 charts which results in some heavy duty math crunching. I think the only reasonable way to slow this down is to group the hourly supply chart on 'Must-run', 'Dispatchable', 'Volatile', etc.

The next culprit in line is the transformation of the data... which is obvious (the downsampling, or the slicing of the load). Since it's just a lot. Maybe I could shave some time of that thing.

**Comment by jorisberkhout** [comment]
> I think the only reasonable way to slow this down is to group the hourly supply chart on 'Must-run', 'Dispatchable', 'Volatile', etc.

See also https://github.com/quintel/etmodel/issues/2106.

**Comment by grdw** [comment]
@antw could you maybe take a look with us to discover some potential speed improvements. 😅

**Comment by antw** [comment]
> The next culprit in line is the transformation of the data... which is obvious (the downsampling, or the slicing of the load). Since it's just a lot. Maybe I could shave some time of that thing.

I feel like a lot of time could be shaved off that thing. :grinning:

![screen shot 2016-05-20 at 13 32 44](https://cloud.githubusercontent.com/assets/4383/15427797/5c395688-1e8f-11e6-9140-fd42c582baeb.png)

As I understand it, the rendering process goes like this:

1. Receive data from server.
2. Transform data into a nicer format (`convertToDateRange`)
3. Slice out the week to be shown, or downsample when viewing the entire year (`MeritTransformator.transform`).
4. Render the chart.

If that is the case, steps 2 and 3 seem are the wrong way around. The data should be received from the server and stored without being transformed (in case the user wants to view a different week), then the weekly data extracted and transformed/formatted. That way we only having to deal with 168 data points per series, instead of 8760. I suspect this is why `D3YearlyChartView.convertToDateRange` and `MeritTransformator.transform` dominate the timeline above. Not only is it slow to work will a full year of data when you only need one week, but it causes quite a bit of GC (the darker yellow blocks at the bottom of the timeline, ~156ms).

1. Receive data from server.
2. Slice out the week to be shown, or downsample when viewing the entire year (`MeritTransformator.transform`).
3. Transform data into a nicer format (`convertToDateRange`)
4. Render the chart.

If the user selects a different week from the drop-down, go to step 2: slice out the new week, transform data, re-render.

**Comment by antw** [comment]
`LoadSlicer.slice` is also slow for this reason. It has to filter over the entire data set for each series, extracting the desired date range. Using `filter` means even if you only want the first week (data points 0 to 168) it calls the filter function over all 8760 elements.

As a quick experiment, I swapped the contents of `LoadSlicer.slice` for:

```js
return serie.oldValues.slice(0, 168);
```

... which reduced the runtime of `MeritTransformator.transform` from 263ms to 2.59ms.

**Comment by grdw** [comment]
It all sounds nice on paper, love to do it. Will cost a long time to implement *today* because of mental reasons. I'll do this, but not now. Right now I'm having a mental breakdown because of `Object`'s not cloning properly and than deciding to a major `$.extend`-inception yadeeyadee - pure Javascript gore at it's finest..... so monday I'll take a look and go : "Ohh, that's how to it" and fix it in 5 minutes instead of waste 3 hours of writing complete hacky-tacky crapcode which I will entertain nobody with. 😅

**Comment by dennisschoenmakers** [comment]
> It all sounds nice on paper, love to do it. Will cost a long time to implement today because of mental reasons. I'll do this, but not now. Right now I'm having a mental breakdown because of Object's not cloning properly and than deciding to a major $.extend-inception yadeeyadee - pure Javascript gore at it's finest..... so monday I'll take a look and go : "Ohh, that's how to it" and fix it in 5 minutes instead of waste 3 hours of writing complete hacky-tacky crapcode which I will entertain nobody with. 😅

tl;dr: Gerard needs a beer! 🍺 We're off to the Fles in an hour! 😄

**Comment by antw** [comment]
>  pure Javascript gore at it's finest

![javascript-the-good-parts-the-definitive-guide](https://cloud.githubusercontent.com/assets/4383/15431689/dfb3a6c2-1ea2-11e6-9ede-313bd1be7a54.jpg)

**Comment by dennisschoenmakers** [comment]
This whole experience is one more reason for me to support pure JS in applications in favor of CoffeeScript

**Comment by antw** [comment]
> This whole experience is one more reason for me to support pure JS in applications in favor of CoffeeScript

[ES2015 plz](https://babeljs.io/docs/learn-es2015/).

**Comment by grdw** [comment]
[You don't know JS!](https://github.com/getify/You-Dont-Know-JS)

**Comment by grdw** [comment]
Current stats:

| Flexibility curve | Hourly supply curve | Price curve |
| -------------------- | --------------------------- | --------------- | 
| ~1500ms           | ~2500ms                  | ~1000ms   |

The only thing that's speed up is the hourly supply chart with this approach. I see no difference whatsoever. The main reason is that there's one `$.extend` needed to copy a lot of data in order to not damage the original imported data. And that method slurps up a lot of time.

**Comment by dennisschoenmakers** [comment]
> Current stats:

> | Flexibility curve | Hourly supply curve | Price curve |
> | -------------------- | --------------------------- | --------------- | 
> | ~1500ms           | ~2500ms                  | ~1000ms   |
 
On which laptop is this measured?

The max, max, max that is allowed would be **1 second** of JS execution time.

**Comment by dennisschoenmakers** [comment]
> The max, max, max that is allowed would be 1 second of JS execution time.

That would be measured on a *median* laptop.

**Comment by antw** [comment]
Here are results from my laptop. Not a "median" machine for sure, but three years old and the results are low enough that I can be confident 1s will not be exceeded on an older laptop:

| Flexibility curve | Hourly supply curve | Price curve |
| -------------------- | --------------------------- | --------------- | 
| ~108ms           | ~189ms                  | ~182ms   |

That is **JS execution time only**; from the response being *received* to the chart being rendered.

The rest of the delay is down to ETEngine. Turning on the Merit order by default came with a big performance penalty. Right now, "normal" ETE requests (requests that don't ask for merit order curves) are taking 1-1.5 seconds; the merit order data for the charts adds a couple of hundred milliseconds to that.

**Comment by dennisschoenmakers** [comment]
> Here are results from my laptop. 

I guess that is with the refactored code? Looks good. Can we deploy this to beta and production soon (@jorisberkhout) ?

**Comment by jorisberkhout** [comment]
> I guess that is with the refactored code? Looks good. Can we deploy this to beta and production soon (@jorisberkhout) ?

I am happy if you are, @dennisschoenmakers 😄 . So if you are, we can deploy this to Beta and Live!

**Comment by grdw** [comment]
> That is JS execution time only; from the response being received to the chart being rendered.

How do you measure that time? I use Firefox but that profiler does not really work as expected... @dennisschoenmakers I counted the full time (so including the time of the request).

**Comment by antw** [comment]
> we can deploy this to Beta and Live!

Done.

I'll close this because I don't think we're going to be able to make any more significant gains. We could split up high-res loads into quarters (Q1, Q2, etc), and down-sample on ETEngine, but we'd likely gain tens of milliseconds at best. That's not nothing, but in the big scheme of things I don't think it's worth the extra queries and complexity in ETE.

> How do you measure that time? I use Firefox but that profiler does not really work as expected...

In the Chrome developer tools using the Timeline. Press Cmd-E to start recording, move a slider and wait for the chart to re-render, and then hit "Finish". I was measuring the total time required to execute the response callback:

![screen shot 2016-05-23 at 15 37 15](https://cloud.githubusercontent.com/assets/4383/15474331/4f636ff6-20fc-11e6-9f27-50d01369f13f.png)

So that measures the time to parse the response JSON, Backbone to update the models (and fire its own callbacks), and (re-)render the chart.

**Comment by dennisschoenmakers** [comment]
Thanks @antw! it is a lot quicker now. I've perhaps spotted another issue, but that should ruin this party 🎈
